File: hydra-lisp/hydra-clojure/src/main/clojure/hydra/lib/regex.clj
(ns hydra.lib.regex
  (:import [java.util.regex Pattern Matcher]))

;; matches :: String -> String -> Bool
(def hydra_lib_regex_matches
  (fn [pattern] (fn [input] (boolean (re-matches (re-pattern pattern) input)))))

;; find :: String -> String -> Maybe String
(def hydra_lib_regex_find
  (fn [pattern] (fn [input]
    (let [m (re-find (re-pattern pattern) input)]
      (if (some? m)
        (if (string? m) m (first m))
        nil)))))

;; find_all :: String -> String -> [String]
(def hydra_lib_regex_find_all
  (fn [pattern] (fn [input]
    (let [results (re-seq (re-pattern pattern) input)]
      (if results
        (map (fn [m] (if (string? m) m (first m))) results)
        ())))))

;; replace :: String -> String -> String -> String
(def hydra_lib_regex_replace
  (fn [pattern] (fn [replacement] (fn [input]
    (let [m (re-matcher (re-pattern pattern) input)]
      (if (.find m)
        (str (.substring input 0 (.start m))
             replacement
             (.substring input (.end m)))
        input))))))

;; replace_all :: String -> String -> String -> String
(def hydra_lib_regex_replace_all
  (fn [pattern] (fn [replacement] (fn [input]
    (.replaceAll (re-matcher (re-pattern pattern) input)
                 (Matcher/quoteReplacement replacement))))))

;; split :: String -> String -> [String]
(def hydra_lib_regex_split
  (fn [pattern] (fn [input]
    (seq (.split (Pattern/compile pattern) input -1)))))

File: hydra-scala/src/main/scala/hydra/lib/regex.scala
package hydra.lib

import java.util.regex.{Matcher, Pattern}

object regex:
  /** Find the first substring matching a regex pattern. */
  def find(pattern: String)(input: String): Option[String] =
    val m = Pattern.compile(pattern).matcher(input)
    if m.find() then Some(m.group()) else None

  /** Find all non-overlapping substrings matching a regex pattern. */
  def findAll(pattern: String)(input: String): Seq[String] =
    val m = Pattern.compile(pattern).matcher(input)
    val results = scala.collection.mutable.ArrayBuffer[String]()
    while m.find() do results += m.group()
    results.toSeq

  /** Check whether an entire string matches a regex pattern. */
  def matches(pattern: String)(input: String): Boolean =
    Pattern.matches(pattern, input)

  /** Replace the first occurrence of a regex pattern with a replacement string. */
  def replace(pattern: String)(replacement: String)(input: String): String =
    Pattern.compile(pattern).matcher(input).replaceFirst(Matcher.quoteReplacement(replacement))

  /** Replace all non-overlapping occurrences of a regex pattern with a replacement string. */
  def replaceAll(pattern: String)(replacement: String)(input: String): String =
    Pattern.compile(pattern).matcher(input).replaceAll(Matcher.quoteReplacement(replacement))

  /** Split a string by a regex pattern. */
  def split(pattern: String)(input: String): Seq[String] =
    Pattern.compile(pattern).split(input, -1).toSeq
